### PR TITLE
handler/PUT: Handle signature verification through a custom reader.

### LINKED ISSUE #2054: Incorrect sha256 should return appropriate error
Incorrect sha256 should return appropriate error:
## Expected behaviour

minio server should return:

```
{ [S3Error: The provided 'x-amz-content-sha256' header does not match what was computed.]
  code: 'XAmzContentSHA256Mismatch',
  clientcomputedcontentsha256: 'a48f6090e6fb65580f31b64209ea90532e0871a55486fee03fe2e6c11965037d',
  s3computedcontentsha256: 'a48f6090e6fb65580f31b64209ea90532e0871a55486fee03fe2e6c11965037c',
  requestid: 'D92E65932905719B',
  hostid: 'xP6PdVqdAFxI3oWqlsnaEsUK459fiCSW+Cr9xXawQSZBJqgwRMdIxgdGO0mb3D+l+5yfPvr1Cz0=',
  amzRequestid: null,
  amzId2: null,
  amzBucketRegion: null }
```
## Actual behaviour

returns invalid signature error

### LINKED ISSUE #2058: Figure out a way to avoid io.Pipe() for PutObjectHandlers

### LINKED ISSUE #2087: XL: Downloading a big file will crash the server
To reproduce: $ mc cp myminio/bucket/5gb.bin /tmp/

This regression is very new, must be due to a commit of today or yesterday

Minio server stack trace:

```
fatal error: runtime: out of memory

runtime stack:
runtime.throw(0xb4a480, 0x16)
    /home/vadmeste/work/go/src/runtime/panic.go:547 +0x90
runtime.sysMap(0xc820f00000, 0x140000000, 0xffffffff0008f800, 0xec4a98)
    /home/vadmeste/work/go/src/runtime/mem_linux.go:206 +0x9b
runtime.(*mheap).sysAlloc(0xeaa840, 0x140000000, 0x1ecf3c2a8)
    /home/vadmeste/work/go/src/runtime/malloc.go:429 +0x191
runtime.(*mheap).grow(0xeaa840, 0xa0000, 0x0)
    /home/vadmeste/work/go/src/runtime/mheap.go:651 +0x63
runtime.(*mheap).allocSpanLocked(0xeaa840, 0xa0000, 0xc81ffb0100)
    /home/vadmeste/work/go/src/runtime/mheap.go:553 +0x4f6
runtime.(*mheap).alloc_m(0xeaa840, 0xa0000, 0xffffff0100000000, 0x7f4ee8e9ee10)
    /home/vadmeste/work/go/src/runtime/mheap.go:437 +0x119
runtime.(*mheap).alloc.func1()
    /home/vadmeste/work/go/src/runtime/mheap.go:502 +0x41
runtime.systemstack(0x7f4ee8e9ee28)
    /home/vadmeste/work/go/src/runtime/asm_amd64.s:307 +0xab
runtime.(*mheap).alloc(0xeaa840, 0xa0000, 0x10100000000, 0x20)
    /home/vadmeste/work/go/src/runtime/mheap.go:503 +0x63
runtime.largeAlloc(0x140000000, 0xc800000001, 0xc820176ac8)
    /home/vadmeste/work/go/src/runtime/malloc.go:766 +0xb3
runtime.mallocgc.func3()
    /home/vadmeste/work/go/src/runtime/malloc.go:664 +0x33
runtime.systemstack(0xc820022000)
    /home/vadmeste/work/go/src/runtime/asm_amd64.s:291 +0x79
runtime.mstart()
    /home/vadmeste/work/go/src/runtime/proc.go:1051

goroutine 82 [running]:
runtime.systemstack_switch()
    /home/vadmeste/work/go/src/runtime/asm_amd64.s:245 fp=0xc820176960 sp=0xc820176958
runtime.mallocgc(0x140000000, 0x925540, 0x1, 0x678270)
    /home/vadmeste/work/go/src/runtime/malloc.go:665 +0x9eb fp=0xc820176a38 sp=0xc820176960
runtime.newarray(0x925540, 0x140000000, 0xc820176ad8)
    /home/vadmeste/work/go/src/runtime/malloc.go:798 +0xc9 fp=0xc820176a78 sp=0xc820176a38
runtime.makeslice(0x910fa0, 0x0, 0x140000000, 0x0, 0x0, 0x0)
    /home/vadmeste/work/go/src/runtime/slice.go:32 +0x165 fp=0xc820176ac8 sp=0xc820176a78
github.com/minio/minio/pkg/objcache.(*Cache).Create(0xc82047e180, 0xc820accf10, 0xe, 0x140000000, 0x0, 0x0, 0x0, 0x0)
    /home/vadmeste/work/gospace/src/github.com/minio/minio/pkg/objcache/objcache.go:103 +0x173 fp=0xc820176b50 sp=0xc820176ac8
main.xlObjects.GetObject(0xc82000a140, 0x10, 0x10, 0xc8200c0a00, 0x10, 0x10, 0x8, 0x8, 0x9, 0x9, ...)
    /home/vadmeste/work/gospace/src/github.com/minio/minio/xl-v1-object.go:125 +0x1194 fp=0xc820177278 sp=0xc820176b50
main.(*xlObjects).GetObject(0xc8201bb880, 0xc8204129a5, 0x6, 0xc8204129ac, 0x7, 0x0, 0x140000000, 0x7f4eecefe628, 0xc8202289c0, 0x0, ...)
    <autogenerated>:253 +0x136 fp=0xc820177338 sp=0xc820177278
main.objectAPIHandlers.GetObjectHandler(0x7f4eecf79f30, 0xc8201bb880, 0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    /home/vadmeste/work/gospace/src/github.com/minio/minio/object-handlers.go:150 +0x693 fp=0xc820177500 sp=0xc820177338
main.(objectAPIHandlers).GetObjectHandler-fm(0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    /home/vadmeste/work/gospace/src/github.com/minio/minio/api-router.go:49 +0x4a fp=0xc820177530 sp=0xc820177500
net/http.HandlerFunc.ServeHTTP(0xc8202b3aa0, 0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    /home/vadmeste/work/go/src/net/http/server.go:1618 +0x3a fp=0xc820177550 sp=0xc820177530
github.com/minio/minio/vendor/github.com/gorilla/mux.(*Router).ServeHTTP(0xc820017270, 0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    /home/vadmeste/work/gospace/src/github.com/minio/minio/vendor/github.com/gorilla/mux/mux.go:100 +0x29e fp=0xc820177668 sp=0xc820177550
main.redirectHandler.ServeHTTP(0x7f4eecefe128, 0xc820017270, 0xacbf80, 0x6, 0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    /home/vadmeste/work/gospace/src/github.com/minio/minio/generic-handlers.go:71 +0x278 fp=0xc8201776e8 sp=0xc820177668
main.(*redirectHandler).ServeHTTP(0xc820294f40, 0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    <autogenerated>:120 +0xb4 fp=0xc820177728 sp=0xc8201776e8
main.minioPrivateBucketHandler.ServeHTTP(0x7f4eecefe150, 0xc820294f40, 0xacbf80, 0x6, 0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    /home/vadmeste/work/gospace/src/github.com/minio/minio/generic-handlers.go:119 +0x1a5 fp=0xc820177790 sp=0xc820177728
main.(*minioPrivateBucketHandler).ServeHTTP(0xc820294f60, 0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    <autogenerated>:122 +0xb4 fp=0xc8201777d0 sp=0xc820177790
main.cacheControlHandler.ServeHTTP(0x7f4eecefe178, 0xc820294f60, 0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    /home/vadmeste/work/gospace/src/github.com/minio/minio/generic-handlers.go:100 +0x26a fp=0xc820177870 sp=0xc8201777d0
main.(*cacheControlHandler).ServeHTTP(0xc82018af90, 0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    <autogenerated>:121 +0xb6 fp=0xc8201778a8 sp=0xc820177870
main.timeHandler.ServeHTTP(0x7f4eecefe1a0, 0xc82018af90, 0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    /home/vadmeste/work/gospace/src/github.com/minio/minio/generic-handlers.go:187 +0x25a fp=0xc820177950 sp=0xc8201778a8
main.(*timeHandler).ServeHTTP(0xc82018afa0, 0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    <autogenerated>:123 +0xb6 fp=0xc820177988 sp=0xc820177950
github.com/minio/minio/vendor/github.com/rs/cors.(*Cors).Handler.func1(0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    /home/vadmeste/work/gospace/src/github.com/minio/minio/vendor/github.com/rs/cors/cors.go:185 +0x1c5 fp=0xc8201779e8 sp=0xc820177988
net/http.HandlerFunc.ServeHTTP(0xc820294fe0, 0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    /home/vadmeste/work/go/src/net/http/server.go:1618 +0x3a fp=0xc820177a08 sp=0xc8201779e8
main.resourceHandler.ServeHTTP(0x7f4eecf7a158, 0xc820294fe0, 0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    /home/vadmeste/work/gospace/src/github.com/minio/minio/generic-handlers.go:247 +0x392 fp=0xc820177aa8 sp=0xc820177a08
main.(*resourceHandler).ServeHTTP(0xc82018b050, 0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    <autogenerated>:124 +0xb6 fp=0xc820177ae0 sp=0xc820177aa8
main.authHandler.ServeHTTP(0x7f4eecefe1f0, 0xc82018b050, 0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    /home/vadmeste/work/gospace/src/github.com/minio/minio/auth-handler.go:169 +0x76 fp=0xc820177b28 sp=0xc820177ae0
main.(*authHandler).ServeHTTP(0xc82018b060, 0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    <autogenerated>:55 +0xb6 fp=0xc820177b60 sp=0xc820177b28
net/http.serverHandler.ServeHTTP(0xc820081d00, 0x7f4eecefe450, 0xc8202289c0, 0xc8202f2d20)
    /home/vadmeste/work/go/src/net/http/server.go:2081 +0x19e fp=0xc820177bc0 sp=0xc820177b60
net/http.(*conn).serve(0xc820080100)
    /home/vadmeste/work/go/src/net/http/server.go:1472 +0xf2e fp=0xc820177f88 sp=0xc820177bc0
runtime.goexit()
    /home/vadmeste/work/go/src/runtime/asm_amd64.s:1998 +0x1 fp=0xc820177f90 sp=0xc820177f88
created by net/http.(*Server).Serve
    /home/vadmeste/work/go/src/net/http/server.go:2137 +0x44e

goroutine 1 [IO wait]:
net.runtime_pollWait(0x7f4eecf79c60, 0x72, 0x0)
    /home/vadmeste/work/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).Wait(0xc82001d800, 0x72, 0x0, 0x0)
    /home/vadmeste/work/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc82001d800, 0x0, 0x0)
    /home/vadmeste/work/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).accept(0xc82001d7a0, 0x0, 0x7f4eecefe2d8, 0xc820412000)
    /home/vadmeste/work/go/src/net/fd_unix.go:426 +0x27c
net.(*TCPListener).AcceptTCP(0xc8202701c8, 0x5100c0, 0x0, 0x0)
    /home/vadmeste/work/go/src/net/tcpsock_posix.go:254 +0x4d
net/http.tcpKeepAliveListener.Accept(0xc8202701c8, 0x0, 0x0, 0x0, 0x0)
    /home/vadmeste/work/go/src/net/http/server.go:2427 +0x41
net/http.(*Server).Serve(0xc820081d00, 0x7f4eecefe2a0, 0xc8202701c8, 0x0, 0x0)
    /home/vadmeste/work/go/src/net/http/server.go:2117 +0x129
net/http.(*Server).ListenAndServe(0xc820081d00, 0x0, 0x0)
    /home/vadmeste/work/go/src/net/http/server.go:2098 +0x136
main.serverMain(0xc82022a000)
    /home/vadmeste/work/gospace/src/github.com/minio/minio/server-main.go:340 +0x1090
github.com/minio/minio/vendor/github.com/minio/cli.Command.Run(0xad55f0, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0xb502c0, 0x1c, 0x0, ...)
    /home/vadmeste/work/gospace/src/github.com/minio/minio/vendor/github.com/minio/cli/command.go:117 +0x19b2
github.com/minio/minio/vendor/github.com/minio/cli.(*App).Run(0xc8200e9560, 0xc82000a120, 0x12, 0x12, 0x0, 0x0)
    /home/vadmeste/work/gospace/src/github.com/minio/minio/vendor/github.com/minio/cli/app.go:186 +0x476
github.com/minio/minio/vendor/github.com/minio/cli.(*App).RunAndExitOnError(0xc8200e9560)
    /home/vadmeste/work/gospace/src/github.com/minio/minio/vendor/github.com/minio/cli/app.go:197 +0x45
main.main()
    /home/vadmeste/work/gospace/src/github.com/minio/minio/main.go:200 +0x1f8

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
    /home/vadmeste/work/go/src/runtime/asm_amd64.s:1998 +0x1

goroutine 5 [syscall]:
os/signal.signal_recv(0x0)
    /home/vadmeste/work/go/src/runtime/sigqueue.go:116 +0x132
os/signal.loop()
    /home/vadmeste/work/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
    /home/vadmeste/work/go/src/os/signal/signal_unix.go:28 +0x37

goroutine 68 [IO wait]:
net.runtime_pollWait(0x7f4eecf79d20, 0x72, 0xc82048c400)
    /home/vadmeste/work/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).Wait(0xc82021a0d0, 0x72, 0x0, 0x0)
    /home/vadmeste/work/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc82021a0d0, 0x0, 0x0)
    /home/vadmeste/work/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc82021a070, 0xc82048c400, 0x400, 0x400, 0x0, 0x7f4eecf74050, 0xc82000e1f0)
    /home/vadmeste/work/go/src/net/fd_unix.go:250 +0x23a
net.(*conn).Read(0xc82021c008, 0xc82048c400, 0x400, 0x400, 0x0, 0x0, 0x0)
    /home/vadmeste/work/go/src/net/net.go:172 +0xe4
crypto/tls.(*block).readFromUntil(0xc820216420, 0x7f4eecf7c2b8, 0xc82021c008, 0x5, 0x0, 0x0)
    /home/vadmeste/work/go/src/crypto/tls/conn.go:460 +0xcc
crypto/tls.(*Conn).readRecord(0xc8201a4900, 0xbf1417, 0x0, 0x0)
    /home/vadmeste/work/go/src/crypto/tls/conn.go:562 +0x2d1
crypto/tls.(*Conn).Read(0xc8201a4900, 0xc820185000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
    /home/vadmeste/work/go/src/crypto/tls/conn.go:939 +0x167
bufio.(*Reader).fill(0xc820226780)
    /home/vadmeste/work/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).Read(0xc820226780, 0xc8201363a8, 0x9, 0x9, 0x9, 0x0, 0x0)
    /home/vadmeste/work/go/src/bufio/bufio.go:207 +0x260
io.ReadAtLeast(0x7f4eecf80768, 0xc820226780, 0xc8201363a8, 0x9, 0x9, 0x9, 0x0, 0x0, 0x0)
    /home/vadmeste/work/go/src/io/io.go:297 +0xe6
io.ReadFull(0x7f4eecf80768, 0xc820226780, 0xc8201363a8, 0x9, 0x9, 0xe2d1a2139f75d01, 0x0, 0x0)
    /home/vadmeste/work/go/src/io/io.go:315 +0x62
net/http.http2readFrameHeader(0xc8201363a8, 0x9, 0x9, 0x7f4eecf80768, 0xc820226780, 0xec000000, 0xc800000000, 0x0, 0x0)
    /home/vadmeste/work/go/src/net/http/h2_bundle.go:715 +0xa5
net/http.(*http2Framer).ReadFrame(0xc820136370, 0x0, 0x0, 0x0, 0x0)
    /home/vadmeste/work/go/src/net/http/h2_bundle.go:906 +0xe9
net/http.(*http2clientConnReadLoop).run(0xc8203de000, 0x0, 0x0)
    /home/vadmeste/work/go/src/net/http/h2_bundle.go:5583 +0xa6
net/http.(*http2ClientConn).readLoop(0xc82007e840)
    /home/vadmeste/work/go/src/net/http/h2_bundle.go:5544 +0x162
created by net/http.(*http2Transport).NewClientConn
    /home/vadmeste/work/go/src/net/http/h2_bundle.go:4975 +0xd94
```

### PR DESCRIPTION
**Pull request by harshavardhana**
Change brings in a new signVerifyReader which provides a io.Reader
compatible reader, additionally implements Verify() function.

Verify() function validates the signature present in the incoming
request. This approach is choosen to avoid complexities involved
in using io.Pipe().

Thanks to Krishna for his inputs on this.

Fixes #2058
Fixes #2054
Fixes #2087

### CODE REVIEW

**Comment by krisis** [comment]
This PR looks good to me. tl;dr

To summarise the problem, the io.Pipe() based approach in current master may fail to detect signature mismatch when the write to disks race the go-routine verifying sha256, especially when there are many concurrent put-object requests.
This is because a `Close()` by the `PipeReader` doesn't return error to the `PipeWriter` and therefore the write to the disks may be persisted despite a sha256 mismatch. So, this fix is necessary for us to ensure correctness.

**Comment by krisis** [comment]
[2066-blockprof.pdf](https://github.com/minio/minio/files/344362/2066-blockprof.pdf)
Attached above is the block profile taken (with this patch on current master) when 200 concurrent uploads/downloads using aws s3 command (which internally uploads parts of an object in parallel).

**Comment by krishnasrinivas** [comment]
> To summarise the problem, the io.Pipe() based approach in current master may fail to detect signature mismatch when the write to disks race the go-routine verifying sha256, especially when there are many concurrent put-object requests.
> This is because a Close() by the PipeReader doesn't return error to the PipeWriter and therefore the write to the disks may be persisted despite a sha256 mismatch. So, this fix is necessary for us to ensure correctness.

I think there won't be any race condition, because if there is a signature mismatch, from the object-layer we call writer.CloseWithError(err) which causes erasureReadFile to return error and finally xl.PutObject returns the error at the object-layer. So we will still be able to catch the error fine.

File: object-handlers.go [hunk 1]
```diff
@@ -600,75 +629,20 @@ func (api objectAPIHandlers) PutObjectHandler(w http.ResponseWriter, r *http.Req
...
-			}()
-
-			// Create object.
-			md5Sum, err = api.ObjectAPI.PutObject(bucket, object, size, reader, metadata)
-			// Close the pipe.
-			reader.Close()
-			// Wait for all the routines to finish.
-			wg.Wait()
+			reader = r.Body
+			signVerifyFn = signVerifyFactory(nil, r)
```
Comment (krishnasrinivas) [inline object-handlers.go]: https would mean we most of the times hit this code path, so even if the signature is wrong we end up reading the entire body. better would be:

```
if err:= signVerifyFactory(nil, r)(); err != nil {
    return
}
```

File: object-interface.go [hunk 2]
```diff
@@ -33,13 +33,13 @@ type ObjectLayer interface {
 	// Object operations.
 	GetObject(bucket, object string, startOffset int64, length int64, writer io.Writer) (err error)
 	GetObjectInfo(bucket, object string) (objInfo ObjectInfo, err error)
-	PutObject(bucket, object string, size int64, data io.Reader, metadata map[string]string) (md5 string, err error)
+	PutObject(bucket, object string, size int64, data io.Reader, metadata map[string]string, signVerify func() error) (md5 string, err error)
```
Comment (krishnasrinivas) [inline object-interface.go]: you've declared `signVerifyFunc`

**Comment by harshavardhana** [comment]
> you've declared signVerifyFunc

ah yes..

File: object-handlers.go (hunk 1 above)
Comment (harshavardhana) [inline object-handlers.go]: > https would mean we most of the times hit this code path, so even if the signature is wrong we end up reading the entire body. better would be:
> 
> if err:= signVerifyFactory(nil, r)(); err != nil {
>     return
> }

Will do this optimization.

**Comment by krisis** [comment]
> I think there won't be any race condition, because if there is a signature mismatch, from the object-layer we call writer.CloseWithError(err) which causes erasureReadFile to return error and finally xl.PutObject returns the error at the object-layer. So we will still be able to catch the error fine.

If the reader calls Close() before the sha256 sum writer detects an error, then the xl.PutObject would have gone ahead and renamed the files representing the object into the namespace.

**Comment by krishnasrinivas** [comment]
> If the reader calls Close() before the sha256 sum writer detects an error, then the xl.PutObject would have gone ahead and renamed the files representing the object into the namespace.

no, see:

```
                // Create object.                                                                               
                md5Sum, err = api.ObjectAPI.PutObject(bucket, object, size, reader, metadata)                   
                // Close the pipe.                                                                              
                reader.Close() 
```

reader.Close() still happens after xl.PutObject
on a side note reader.Close() is not needed here i guess as it does not serve any purpose.

**Comment by harshavardhana** [comment]
> I think there won't be any race condition, because if there is a signature mismatch, from the object-layer we call writer.CloseWithError(err) which causes erasureReadFile to return error and finally xl.PutObject returns the error at the object-layer. So we will still be able to catch the error fine.

This is only theory, but we saw an odd behavior perhaps even found a bug in io.Pipe() -  in case of a large load what really happens is that the go routine never returns a signature mismatch by the time it has returned the other end got success and wrote the file fully. 

Since the signature error has a poor way of indicating using CloseWithError() there is a possibility of missing this error, and we can readily reproduce this without the change #2059 

There is one classic case already look at the TestPutObjectLongName test cases in server_test.go and see the 'buffer'

**Comment by krisis** [comment]
> reader.Close() still happens after xl.PutObject
> on a side note reader.Close() is not needed here i guess as it does not serve any purpose.

Yes, but could happen before the sha256 sum writer completes verification of the body. This means that the writes to the disk could complete before the signature verification is over.
The point is that the reader, i.e xl.PutObjectPart routine could complete before the go-routine performing the signature verification. This means there would be times when there might be a signature mismatch, yet we would have written the part into the namespace.

**Comment by krisis** [comment]
[2066-cpuprof.pdf](https://github.com/minio/minio/files/344445/2066-cpuprof.pdf)
Attached above is the cpu profile for the same workload.

**Comment by harshavardhana** [comment]
In-fact @krishnasrinivas is right, its just that current master doesn't cater for situations where ContentLength can be set to '-1'  and can lead to mis-behavior this is indeed what we saw.

This patch caters for even '-1' situation and returns proper error. A test case is added to indicate that "TestValidateSignature"

**Pull request merged**